### ISSUE TITLE
Supporting `edge_softmax` for `dgl.heterograph`

### ISSUE THREAD

**Issue by Junyoungpark** [dmlc/dgl#2446]
## 🚀 Feature
Hi, while I working with `dgl.heterograph`, I found that there's no equivalent method `edge_softmax`  for `dgl.heterograph`. I already check [this thread](https://discuss.dgl.ai/t/edge-softmax-on-dglheterograph/834/4) as a workaround for the `edge_softmax` on `dgl.heterograph`. In short,
(1) convert `dgl.heterograph` into `dgl.graph` and (2) do `edge_softmax` (3) Convert the `dgl.graph` back to the `dgl.heterograph`.

## Motivation

I wonder how much the computational overhead (memory or speed) of the proposed method and also want to know what if we use the following code to do `edge_softmax` on `dgl.heterograph`

## Alternatives

```py
  
import dgl
import torch


def hetero_edge_softmax(g, etype,
                        logit_from: str,
                        logit_field='z',
                        attn_field='attn'):
    with g.local_scope():
        g.edges[etype].data[logit_field] = torch.exp(g.edges[etype].data[logit_from])
        g.update_all(dgl.function.copy_e(logit_field, logit_field),
                     dgl.function.sum(logit_field, 'sum_logit'),
                     etype=etype)

        def _compute_sm(edges):
            summed = edges.dst['sum_logit']
            logit = edges.data[logit_field]
            return {attn_field: logit / summed}

        g.apply_edges(_compute_sm, etype=etype)
        return g.edges[etype].data[attn_field]
```

**Comment by yzh119** [comment]
@Junyoungpark thanks for the proposal, actually I think what you did in the your function is equivalent to:
```python
dgl.ops.edge_softmax(g[etype], edata)
```
and there is no need to implement another `hetero_edge_softmax` function.

**Comment by yzh119** [comment]
[Thread 834](https://discuss.dgl.ai/t/edge-softmax-on-dglheterograph/834/4) was focusing more on cross edge type softmax, and I don't think your code snippet handle the "cross edge type" issue.

**Comment by yzh119** [comment]
Some comments:
```python
        def _compute_sm(edges):
            summed = edges.dst['sum_logit']
            logit = edges.data[logit_field]
            return {attn_field: logit / summed}

        g.apply_edges(_compute_sm, etype=etype)
```
is equivalent to
```
g.apply_edges(dgl.function.e_div_v(logit_field, 'sum_logit', attn_field))
```
while the builtin function is much more efficient.

Another problem is numerical issue, `exp` is easily to overflow, a common approach to this is to minus the logits by its maximum: https://stackoverflow.com/questions/42599498/numercially-stable-softmax, and this is exactly what dgl's edge softmax did:

https://github.com/dmlc/dgl/blob/master/python/dgl/backend/pytorch/sparse.py#L189-L194

**Comment by Junyoungpark** [comment]
Thanks for the detailed explanation. I would need to change my code also to use the way as you suggested.
